community/csv/src/main/java/org/neo4j/csv/reader/ProcessingSource.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.csv.reader;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+
+import org.neo4j.csv.reader.Source.Chunk;
+
+import static java.lang.Math.min;
+
+/**
+ * In a scenario where there's one reader reading chunks of data, handing those chunks to one or
+ * more processors (parsers) of that data, this class comes in handy. This pattern allows for
+ * multiple {@link BufferedCharSeeker seeker instances}, each operating over one chunk, not transitioning itself
+ * into the next.
+ */
+public class ProcessingSource implements Closeable
+{
+    private final CharReadable reader;
+    private final int chunkSize;
+    private char[] backBuffer; // grows on demand
+    private int backBufferCursor;
+
+    // Buffer reuse, grows on demand up to number of processors concurrently parsing character data
+    private final int maxNumberOfBufferedChunks;
+    private volatile char[][] buffers = new char[3][];
```

Comment:
The `volatile` indications that this class will face concurrent access, but I don't see any other form of synchronisation going on in here.